Implement dimming for sliding on sliders with flashlight mod

nyquillerium opened the pull request:
* Closes https://github.com/ppy/osu/issues/4238
* [x] Depends on https://github.com/ppy/osu-resources/pull/53

Test with https://ci.appveyor.com/project/peppy/osu-resources/builds/23790165
Package: osu.Game.Resources
Feed: https://ci.appveyor.com/nuget/osu-resources-f7frosjy492d
Version: 11.0.0

There's a few things I'm uncertain of:
 * ~~I matched the dim value with stable, but it still seems a little more _jarring_ than how it is there. Not sure if its just me. Wondering if its worth doing a really fast fade instead of an instant toggle for the dim.~~ Added a fade duration.
 * ~~I tested this with a version of TestCaseOsuPlayer, but there's a black box there so I guess it's not meant for visual testing at this time?~~ Fixed the test cases. 
* ~~There seems to be a separate issue with flashlight size if I try to test it there (its too big, related to https://github.com/ppy/osu/issues/3740 maybe)~~. Fixed the sizing in ModFlashlight to only use the scale of the playfield.
 * The starting flashlight position is wrong, but that also happens on master and is probably related to https://github.com/ppy/osu/issues/3740.

Also fixes a problem with player test cases not being visible (now they all by default have 1.0 dim).

nyquillerium committed:
Implement flashlight dimming on slider slide

nyquillerium committed:
Invert flashlight dim

peppy commented:
I would hope we can add a brief fade (50ms should work well).

As for the black box issues, can you post screenshots of what you're seeing and how you're testing (code wide)? Why not make a flashlight test case in the process?

nyquillerium commented:
I'm referring to the 
```
Add(new Box
            {
                RelativeSizeAxes = Axes.Both,
                Colour = Color4.Black,
                Depth = int.MaxValue
            });
```
added in every player test case BDL, but I guess this is some sort of optimization for CI runs? I removed it for the sake of visual testing, but I reverted it when making this pr. Could potentially add it back as a TestCaseFlashlight with some more code test coverage.

nyquillerium committed:
Add back flashlight testcase

nyquillerium committed:
Rename to TestCaseOsuFlashlight

nyquillerium committed:
Add short fade to flashlight dimming

nyquillerium commented:
found out there's a couple of more edge cases to test, blocking while I do that.

Also, the issue with flashlight size mentioned in the pr description can now be visually seen in TestCaseOsuFlashlight

nyquillerium committed:
Use absolute sequence for flashlight breaks

peppy committed:
Merge remote-tracking branch 'upstream/master' into flashlight-dim

peppy committed:
Update resources and mods usage

peppy commented:
I can't see the flashlight in the testcase, only the dim. Is that intentional?

nyquillerium commented:
The flashlight is there, but the size is just completely wrong in the test case (mentioned in PR description). It looks correct in gameplay, so I assume there's something else going on in flashlight calculations related to the issue linked there.

peppy reviewed (changes requested):
As mentioned, please make test work properly first.

nyquillerium committed:
Use the actual scale of Flashlight for FlashlightSize

nyquillerium committed:
Remove unused using

nyquillerium commented:
I corrected ModFlashlight to extract the scale of the draw matrix instead of using Transform on the draw matrix, and it seems to have resolved everything nicely without changing how the flashlight size behaves on the game side.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModFlashlight.cs:
```diff
@@ -1,31 +1,58 @@
+                    trackingSliders--;
+
+                // If there are any sliders in a tracking state, apply a dim to the entire playfield over a brief duration.
+                this.TransformTo(nameof(FlashlightDim), trackingSliders > 0 ? 0.8f : 0.0f, 50);
```
this should not run unless the count has changed to 0 or 1, for correctness. else you will interrupt an existing transition with a new duration.

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModFlashlight.cs:
```diff
@@ -1,31 +1,58 @@
+
+        public void ApplyToDrawableHitObjects(IEnumerable<DrawableHitObject> drawables)
+        {
+            foreach (DrawableSlider drawable in drawables.OfType<DrawableSlider>())
```
`var slider` or `var s`

peppy commented on osu.Game/Rulesets/Mods/ModFlashlight.cs:
```diff
@@ -94,14 +95,17 @@ protected override void LoadComplete()
-                this.FadeInFromZero(FLASHLIGHT_FADE_DURATION);
-
-                foreach (var breakPeriod in Breaks)
+                using (BeginAbsoluteSequence(0))
```
Might want to check if this is how stable does it. It means the flashlight will nto be visible during lead-in time.

nyquillerium replied:
I could leave this change for a separate PR, currently flashlight fades on master are incorrectly accounting for break time and this change attempts to fix that (see Disco Prince with flashlight as an example). For the entire intro sequence to match stable, it needs a little more work. On stable it does a bit of an expand and then contract, which lazer currently does not do. I'll probably open/edit the current issue with a list of things that flashlight needs to match stable.

The flashlight is visible during the lead-in time because a FadeInFromZero is applied immediately on LoadComplete above this change. There's a separate issue where it attempts to re-apply the flashlight when gameplay starts though. Probably even more reason to make this a different PR.

peppy replied:
I don't follow your response here. You say it was applied "immediately" but I cannot see where that was being done (before you made changes after my review). Please explain what you mean.

nyquillerium replied:
I'm referring to the dim applied on https://github.com/ppy/osu/blob/b2b7143ed328a6d653ed3ce1fb3cdd34594e6f4c/osu.Game/Rulesets/Mods/ModFlashlight.cs#L98

peppy replied:
you mean the one that was moved after my comment..?

nyquillerium replied:
yeah, as I mentioned in the above comment, that was incorrect.

nyquillerium committed:
Use var, rework dim application logic

nyquillerium committed:
Move intiial fade outside of absolute sequence

nyquillerium committed:
Merge branch 'master' into flashlight-dim

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModFlashlight.cs:
```diff
@@ -1,31 +1,74 @@
+                if (e.NewValue)
+                {
+                    trackingSliders++;
+                    // This check being here ensures we're only applying a dim if and only if a slider begins tracking.
```
you mean "only applying a dim if the *first* slider begins tracking", right? the comment makes little sense right now.

nyquillerium replied:
The reason why this logic is specifically here where the new value is true is because we probably don't want to be applying a new transition in the case where the number of tracking sliders goes from 2 to 1 (when the new value is false)

peppy replied:
i was referring to the comment, specifically

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModFlashlight.cs:
```diff
@@ -1,31 +1,74 @@
+            public void OnSliderTrackingChange(ValueChangedEvent<bool> e)
+            {
+                // If any sliders are in a tracking state, apply a dim to the entire playfield over a brief duration.
+                if (e.NewValue)
```
this whole mthod is logically weird.

```csharp
trackingSliders += (e.NewValue ? 1 : -1);

if (trackingSliders < 0)
    throw new InvalidOperationException($"The number of {nameof(trackingSliders)} cannot be below 0.");

switch (trackingSliders)
{
    case 0:
		// do thing
	case 1:
		// do thing
}

```

peppy replied:
^ not valid, but better commenting is required to explain why this cannot work

nyquillerium committed:
Remove fade in, update comment

peppy committed:
Reword and reoganise logic

peppy commented:
The logic looks okay, but how are you testing the case of more than one slider?

peppy committed:
Merge branch 'master' into flashlight-dim

nyquillerium commented:
Upon further testing on stable side, it turns out its impossible to track both sliders in the first place so that logic was unnecessary,

nyquillerium committed:
Remove unnecessary multi-slider tracking logic

nyquillerium committed:
Remove unused usings

nyquillerium committed:
Merge branch 'master' into flashlight-dim

nyquillerium committed:
Fix framework version

peppy commented:
Needs to be updated with new draw node changes.

nyquillerium committed:
Merge remote-tracking branch 'upstream/master' into flashlight-dim

nyquillerium committed:
Fix merge

smoogipoo committed:
Merge remote-tracking branch 'Joehuu/master' into flashlight-dim

smoogipoo committed:
Merge branch 'master' into flashlight-dim
